R/DBConnection.R

```diff
@@ -106,6 +106,103 @@ setGeneric("dbSendQuery",
+#' \code{INSERT INTO}, \code{DROP TABLE}, ...). It will execute
+#' the query and return the number of rows affected by the operation.
+#'
+#' @inheritParams dbDisconnect
+#' @param statement a character vector of length 1 containing SQL.
+#' @return The number of rows affected by the \code{statement}
+#' @aliases dbExecQuery,DBIConnection,character-method
+#' @family connection methods
+#' @export
+#' @examples
+#' con <- dbConnect(RSQLite::SQLite(), ":memory:")
+#' dbWriteTable(con, "cars", head(cars, 3))
+#' dbReadTable(con, "cars")   # there's 3 rows!
+#' dbExecQuery(con, "INSERT INTO cars (speed, dist)
+#'                   VALUES (1, 1), (2, 2), (3, 3);")
+#' dbReadTable(con, "cars")   # there's now 6 rows!
+#' dbDisconnect(con)
+setGeneric("dbExecQuery",
+  def = function(conn, statement, ...) standardGeneric("dbExecQuery")
+)
+
+#' @export
+setMethod("dbExecQuery", signature("DBIConnection", "character"),
+  function(conn, statement, ...) {
+    if (grepl("select", tolower(statement))) {
+      stop("Only use this function for non-SELECT SQL statements.")
+    }
+    rs <- dbSendQuery(conn, statement, ...)
+    on.exit(dbClearResult(rs))
+
+    ## do we need this check in this case?
+    if (!dbHasCompleted(rs)) {
+      warning("Pending rows", call. = FALSE)
+    }
+
+    dbGetRowsAffected(rs)
+  }
+)
+
+#' Use a callback to access the result of a query
+#'
+#' This function allows you to get the result of a query according
+#' to a callback you specify. This is meant as a more efficient and
+#' more general alternative to \code{\link{dbSendQuery}}. On one
+#' hand, it does not keep the result set open, so you don't have to
+#' worry about \code{\link{dbClearResult}}. On the other hand, you
+#' can fetch the result according to your own liking, not only by
+#' fetching an x number of consecutive rows.
+#'
+#' @inheritParams dbDisconnect
+#' @param statement a character vector of length 1 containing SQL.
+#' @param callback a callback function whose first argument must be
+#'  the result set returned by the \code{statement}.
+#' @return A function that accepts the same arguments as the
+#'  \code{callback} (except the first one, which the result set).
+#' @aliases dbGetChunkedQuery,DBIConnection,character-method
+#' @family connection methods
+#' @export
+#' @examples
+#' con <- dbConnect(RSQLite::SQLite(), ":memory:")
+#' dbWriteTable(con, "cars", cars)
+#' res <- dbGetChunkedQuery(con, "SELECT * FROM cars", `[`)
+#' res(1:3,)   # get the first 3 rows
+#' res(15,)    # get the 15th row
+#' dbDisconnect(con)
+setGeneric("dbGetChunkedQuery",
+  def = function(conn, statement, callback, ...) standardGeneric("dbGetChunkedQuery")
+)
+
+#' @export
+setMethod("dbGetChunkedQuery", signature("DBIConnection", "character"),
+  function(conn, statement, callback, ...) {
+    rs <- dbSendQuery(conn, statement, ...)
+    on.exit(dbClearResult(rs))
+
+    df <- tryCatch(
+      dbFetch(rs, n = -1, ...),
```

Comment:
@hadley: Is the `tryCatch()` really unnecessary? @jcheng5 has some strong feelings about this -- that we should let things just error out, rather than catching the error and turning it into a warning... Is there a reason why the current `dbGetQuery()` code uses this structure?